src/common/events.ts

```diff
@@ -0,0 +1,294 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { EventEmitter } from 'eventemitter3';
+import { PassThrough, Readable } from 'stream';
+
+import {
+  TurboSigningEmitterEventArgs,
+  TurboSigningEventsAndPayloads,
+  TurboUploadEmitterEventArgs,
+  TurboUploadEventsAndPayloads,
+} from '../types.js';
+
+/**
+ * Creates an eventing ReadableStream that emits progress and error events using the event names map.
+ *
+ * E.g.
+ *
+ * ```ts
+ * const eventNamesMap = {
+ *   'on-progress': 'signing-progress', // emits 'signing-progress' on event progress
+ *   'on-error': 'signing-error', // emits 'signing-error' errors
+ * };
+ *
+ * const streamWithEvents = createStreamWithEvents({
+ *   data,
+ *   dataSize,
+ *   emitter,
+ *   eventNamesMap,
+ * });
+ * ```
+ */
+function createReadableStreamWithEvents({
+  data,
+  dataSize,
+  emitter,
+  eventNamesMap,
+}: {
+  data: Buffer | ReadableStream;
+  dataSize: number;
+  emitter: EventEmitter;
+  eventNamesMap: {
+    'on-progress': string;
+    'on-error': string;
+  };
+}): ReadableStream {
+  const originalStream =
+    data instanceof ReadableStream
+      ? data
+      : new ReadableStream({
+          start: (controller) => {
+            controller.enqueue(data);
+            controller.close();
+          },
+        });
+
+  let processedBytes = 0;
+  let reader;
+  return new ReadableStream({
+    start() {
+      reader = originalStream.getReader();
+    },
+    async pull(controller) {
+      try {
+        const { value, done } = await reader.read();
+
+        if (done) {
+          emitter.emit(eventNamesMap['on-end']);
+          controller.close();
+          return;
+        }
+
+        processedBytes += value.length;
+        emitter.emit(eventNamesMap['on-progress'], {
+          processedBytes,
+          totalBytes: dataSize,
+        });
+
+        controller.enqueue(value);
+      } catch (error) {
+        emitter.emit(eventNamesMap['on-error'], error);
+        controller.error(error);
+      }
+    },
+    cancel(reason) {
+      return reader.cancel(reason);
+    },
+  });
+}
+
+/**
+ * Creates an eventing Readable stream that emits progress and error events
+ */
+function createReadableWithEvents({
+  data,
+  dataSize,
+  emitter,
+  eventNamesMap,
+}: {
+  data: Readable | Buffer;
+  dataSize: number;
+  emitter: EventEmitter;
+  eventNamesMap: {
+    'on-progress': string;
+    'on-error': string;
+    'on-end': string;
```

Comment:
has extra event ReadableStream doesnt have - see comment re having a type for these.